client/ethmulticlient.go

```diff
@@ -0,0 +1,697 @@
+		return c.SendTransaction(ctx, tx)
+	})
+}
+
+// ReorderClients reorders clients to use in BC calls by the given address slice.
+// If there are a lot of clients this function could be expensive and it will block all other
+// calls or usage of these clients while reordering.
+func (c *EthMultiClient) ReorderClients(addresses []string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if len(c.clients) != len(addresses) {
+		return fmt.Errorf("can't reorder: given %d addresses to use when reordering but have %d clients", len(addresses), len(c.clients))
+	}
+
+	newClients := make([]EthClientGetter, len(c.clients))
+	for i, addr := range addresses {
+		found := false
+		for _, cl := range c.clients {
+			if addr == cl.Address() {
+				newClients[i] = cl
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return fmt.Errorf("can't re-order: no client with address: %s", addr)
+		}
+	}
+
+	c.clients = newClients
+	return nil
+}
+
+// CurrentClientOrder returns the current orders of all clients the multiclient uses.
+func (c *EthMultiClient) CurrentClientOrder() []string {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	result := make([]string, len(c.clients))
+	for i, c := range c.clients {
+		result[i] = c.Address()
+	}
+	return result
+}
+
+// doWithMultipleClientsCtx will execute a given function with all clients.
+//
+// If parent context is cancel or receives a timeout, all calls will be also cancels and the
+// function will return.
+func (c *EthMultiClient) doWithMultipleClients(ctx context.Context, do func(ctx context.Context, c EtherClient) error) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	ctxs, cancel := c.produceCtxs(ctx)
+	defer cancel()
+
+	done := make(chan struct{}, 0)
+	defer close(done)
+
+	go func() {
+		select {
+		case <-done:
+			return
+		case <-ctx.Done():
+			cancel()
+		}
+	}()
+
+	var err error
+	for i, cl := range c.clients {
+		select {
+		case <-ctx.Done():
+			return context.DeadlineExceeded
+		default:
+			childCtx := ctxs[i]
+			if childCtx.Err() != nil && ctx.Err() != nil {
+				return ctx.Err()
+			}
+
+			err = do(childCtx, cl.Client())
+			if err != nil {
+				if errors.Is(err, context.DeadlineExceeded) {
+					// If parent context wasn't canceled that means that we've timed out on this node.
+					if ctx.Err() == nil {
+						c.notifyDowntime(cl.Address())
+					}
+					continue
+				}
+				return err
+			}
+
+			return nil
+		}
+	}
+
+	return err
+}
+
+func (c *EthMultiClient) notifyDowntime(address string) {
+	// If no channel is given, just return.
+	if c.notifyDown == nil {
+		return
+	}
+
+	// If channel is full, drop the notification.
+	select {
+	case c.notifyDown <- address:
+	default:
+	}
+}
+
+func (c *EthMultiClient) closeNotify() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.notifyDown != nil {
+		close(c.notifyDown)
+		// Set equal to `nil` so that any hanging queries dont
+		// get broadcasted as failure on client close.
+		c.notifyDown = nil
```

Comment:
i can see indefinite locks happening here. 

Scenario:

call closeNotify() and notifyDowntime() in multiple goroutines.

You'll run into a situation where c.notifyDown will beome nil after passing the c.notifyDown == nil check, meaning that the following select will block forever.


This needs to be addressed.

Reply:
nice catch